src/uct/base/uct_iface.c

```diff
@@ -219,11 +219,67 @@ int uct_iface_is_reachable(const uct_iface_h iface, const uct_device_addr_t *dev
     return iface->ops.iface_is_reachable(iface, dev_addr, iface_addr);
 }
 
-int uct_iface_is_reachable_v2(const uct_iface_h iface,
-                              const uct_iface_is_reachable_params_t *params)
+static int
+uct_iface_is_intra_node_reachable(const uct_iface_h iface,
+                                  const uct_iface_is_reachable_params_t *params)
 {
-    ucs_fatal("uct_iface_is_reachable_v2 not supported yet");
-    return 0;
+    void *dev_addr;
+    uct_iface_attr_t attr;
+    ucs_status_t status;
+    int reachable;
+
+    status = uct_iface_query(iface, &attr);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    dev_addr = ucs_malloc(attr.device_addr_len, "uct_tmp_dev_addr");
+    if (dev_addr == NULL) {
+        ucs_error("failed to allocated device address of size %zu",
+                  attr.device_addr_len);
+        return 0;
+    }
+
+    status = uct_iface_get_device_address(iface, dev_addr);
+    if (status != UCS_OK) {
+        ucs_free(dev_addr);
+        return 0;
+    }
+
+    reachable = !memcmp(params->device_addr, dev_addr, attr.device_addr_len);
+    ucs_free(dev_addr);
+    return reachable;
+}
+
+static int
+uct_base_iface_is_reachable_v2(const uct_iface_h iface,
+                               uct_iface_is_reachable_params_t *params)
+{
+    params->info_string_length = 0;
+
+    if (params->field_mask & UCT_IFACE_IS_REACHABLE_FIELD_INFO_STRING) {
+        params->info_string[0] = '\0';
+    }
+
+    if ((params->field_mask & UCT_IFACE_IS_REACHABLE_FIELD_SCOPE) &&
+        (params->scope == UCT_IFACE_REACHABILITY_INTRA_NODE)) {
+        return uct_iface_is_intra_node_reachable(iface, params);
+    }
+
+    return uct_iface_is_reachable(iface, params->device_addr,
+                                  params->iface_addr);
+}
+
+int uct_iface_is_reachable_v2(const uct_iface_h tl_iface,
+                              uct_iface_is_reachable_params_t *params)
+{
+    uct_base_iface_t *iface = ucs_derived_of(tl_iface, uct_base_iface_t);
+
+    if (iface->internal_ops->iface_is_reachable_v2 == NULL) {
```

Comment:
all transports should define uct_base_iface_is_reachable_v2 function pointer instead of NULL

Reply:
can you explain why is it better?